HYDRATOR-1323 introduce error transform

albertshau opened the pull request:
An ErrorTransform can be placed after another stage to consume
the errors emitted by that stage. MapReduce changes boil down to
differentiating between stages that takes output and stages that
take errors. When emit() is called, it will only be sent to stages
that take output. When emitError() is called, it will only be sent
to stages that take errors.

Spark refactoring amounts to making stage return a Tuple2 of
boolean and object. The boolean indicates whether the second
object is an error record or an output record. Each stage that
can emit both is then filtered to make two RDDs, one that only
has outputs and one that only has errors.

albertshau committed:
HYDRATOR-1323 introduce error transform

An ErrorTransform can be placed after another stage to consume
the errors emitted by that stage. MapReduce changes boil down to
differentiating between stages that takes output and stages that
take errors. When emit() is called, it will only be sent to stages
that take output. When emitError() is called, it will only be sent
to stages that take errors.

Spark refactoring amounts to making stage return a Tuple2 of
boolean and object. The boolean indicates whether the second
object is an error record or an output record. Each stage that
can emit both is then filtered to make two RDDs, one that only
has outputs and one that only has errors.

albertshau commented:
http://builds.cask.co/browse/CDAP-RUT575

albertshau commented:
just realized schemas are not handled correctly, so putting on hold

anew commented on cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/ErrorRecord.java:
```diff
@@ -0,0 +1,49 @@
+  /**
+   * @return the error code if one was give
+   */
+  @Nullable
```
how can this be null? If I emit an InvalidEntry<T>, its constructor requires that a pass in an error code and error message. So that can never be null?

albertshau replied:
ah yes you're right, not sure why I thought this could be null

anew commented on cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Emitter.java:
```diff
@@ -33,7 +33,8 @@
   /**
-   * Emit an Error object.
+   * Emit an Error object. If an {@link ErrorTransform} is placed after this stage, it will be able to consume
+   * the errors. Otherwise the errors are simply dropped.
```
So the type of the invalid entry must be the same as the type of the regular output objects? Is that practical? 

For example, if I have a transform that converts type A to type B, then it would emit B values. But for some inputs, it may not be possible to convert to type B. Wouldn't the transform want to emit the original input of type A as an invalid entry in this case?

albertshau replied:
it isn't, but I'm not sure how to change this without breaking backwards compatibility. Open to suggestions.

From a practical standpoint, it doesn't matter for any of our plugins since they all operate on StructuredRecord

anew replied:
ah, ok, then let's leave it the way it is.

anew commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformEmitter.java:
```diff
@@ -31,31 +31,34 @@
+      pipeTransformDetail.process(new KeyValue<>(stageName, (Object) errorRecord));
+    }
     try {
-      if (errorOutputWriter == null) {
```
would you still want to warn if there are no errorConsumers?

albertshau replied:
hm maybe, though probably not once per error. I'll change it to log once.

anew commented on cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/BasicErrorRecord.java:
```diff
@@ -0,0 +1,64 @@
+  private final Integer errorCode;
+  private final String errorMessage;
+
+  public BasicErrorRecord(T record, String stageName, @Nullable Integer errorCode, @Nullable String errorMessage) {
```
related to my question above: InvalidEntry takes an int and a non-nullable message, and this seems to be constructed only form InvalidEntry's. So why are the error code and message nullable here?

anew commented on cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/NoErrorEmitter.java:
```diff
@@ -0,0 +1,61 @@
+  private final List<T> entryList;
+  private final String errorMessage;
+
+  public NoErrorEmitter() {
```
Seems like a very specific message for a general-purpose class.

anew commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/batch/RDDCollection.java:
```diff
@@ -78,12 +79,18 @@ public RDDCollection(JavaSparkExecutionContext sec, JavaSparkContext jsc,
 
   @Override
-  public <U> SparkCollection<U> aggregate(StageInfo stageInfo, @Nullable Integer partitions) {
+  public SparkCollection<Tuple2<Boolean, Object>> aggregate(StageInfo stageInfo, @Nullable Integer partitions) {
```
why is not a generic anymore? Can't it be <Tuple2<Boolean, U>>?

albertshau replied:
it can't because it will sometimes be `U` and sometimes be `ErrorRecord<U>`

anew replied:
Oh I see.

anew commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/ErrorFilter.java:
```diff
@@ -0,0 +1,41 @@
+
+  @Override
+  public Iterable<T> call(Tuple2<Boolean, Object> input) throws Exception {
+    List<T> output = new ArrayList<>();
```
I wonder if using Collections.emptyList() and Collections.singletonList() would be more efficient?

anew commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/OutputFilter.java:
```diff
@@ -0,0 +1,42 @@
+  @Override
+  public Iterable<ErrorRecord<T>> call(Tuple2<Boolean, Object> input) throws Exception {
+    List<ErrorRecord<T>> output = new ArrayList<>();
+    if (input._1()) {
```
same as for ErrorFilter

anew commented on cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/transform/DropNullTransform.java:
```diff
@@ -0,0 +1,111 @@
+      String fieldName = field.getName();
+      Object val = input.get(fieldName);
+      if (fieldName.equals(config.field)) {
+        isNotNull = val != null;
```
could just emit the error and return here?

anew commented on cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineRunner.java:
```diff
@@ -87,9 +89,10 @@ public SparkStreamingPipelineRunner(JavaSparkExecutionContext sec, JavaStreaming
-                                                                       "records.out", dataTracer));
-    return new DStreamCollection<>(sec, javaDStream);
+    JavaDStream<Tuple2<Boolean, Object>> outputDStream = javaDStream
+      .transform(new CountingTranformFunction<>(stageInfo.getName(), sec.getMetrics(), "records.out", dataTracer))
```
unrelated: typo in class name... CountingTransformFunction

anew reviewed (commented):
Hi @albertshau , I did my best to review this, left a few comments. Some may be due to my limited context. Thanks.

albertshau committed:
HYDRATOR-1323 error transform code cleanup

Fixing bugs where some fields that are never null were being
marked nullable, and various other small improvements.

albertshau commented:
thanks for the review. I've addressed the comments except for the one about the InvalidEntry type. I'm not sure how to go about addressing it, but I've opened https://issues.cask.co/browse/HYDRATOR-1390 to track it.

anew commented:
LGTM
